Add ability for `instructlab-knowledge` notebook to take multiple source and qna files

alimaredia opened the pull request:
`instructlab-knowledge.ipynb` only accepted one PDF file and created one `qna.yaml` file. This PR allows multiple .pdf files to be converted and chunked and allows multiple knowledge contributions each of which can have multiple source pdf files and have one qna.yaml generated.

This PR also contains minor cleanup also a changing to chunking code from having chunks in individual .txt files to all of the chunks being moved into a JSONL file.

anastasds commented:
What's the design goal behind the new `contributions` idea? There's a lot of newly introduced magic strings that have come in with it.

Ryfernandes reviewed (commented):
Everything ran correctly, and the documentation was clear enough for me to follow the contributions system. A couple of thoughts, although not sure if these should be addressed here or as separate features in separate pr's:

- Since we are using os.environ() for the authoring section, it could be helpful to provide and reference an example.env file
- I have mentioned this before/have built it into my UI, but when users upload multiple files, it would be useful to have multiple pipelines with conversion settings for Docling and to be able to assign them to different documents. For example here, the research paper could benefit from image classification/description and formula enrichment, whereas these are largely unnecessary/a waste of compute for the NFL rulebook. I have a [notebook here](https://github.com/Ryfernandes/docling-table-omission) of a tool I wrote that starts experimenting with some methods of doing this (just specifying conversion settings and adding aliases, not assigning them yet).

iamemilio commented:
I agree with Ryan about wanting different conversion options based on document layout and composition.

I am starting to feel like this is doing too much, and I am not really sure why we are forcing it to be in a notebook, when we're really writing a data conversion pipeline tool on top of docling, with input and output data. If that's what we are delivering, wouldn't it be better off as a product, with standards, testing, explicit inputs and outputs, an sdk/cli/ui... etc, rather than an increasingly complex script that you run in a notebook?

khaledsulayman commented:
> I am starting to wonder if the `instructlab-knoweldge` notebook is doing too much, and I am not really sure why we are forcing it to be in a notebook, when it seems like what we want to do is to write a data conversion pipeline tool that wraps docling. If that's what we are delivering, wouldn't it be better off as a product, with standards, testing, explicit inputs and outputs, an sdk/cli/ui... etc, rather than an increasingly complex script that you run in a notebook?

@iamemilio I see where you're coming from, however I would lean against drawing boxes around any particular part of the workflow at the moment. Modularization is definitely the ultimate goal for a product, but from my understanding the initial push towards notebooks was so they could serve as a bucket of code where we could get an unbiased sense of all the customizability we would want. I think it makes sense to build this up to include all the customizability we would want and then break it down into more concrete modules once product requirements are clearer.

kelbrown20 commented on notebooks/instructlab-knowledge/instructlab-knowledge.ipynb:
```diff
@@ -330,323 +510,128 @@
-   "metadata": {},
-   "source": [
-    "### Define metadata for qna.yaml"
+    "Before and after any editting is done, the `qna.yaml` files can be quickly reviewed to ensure they have the proper elements and number of elements. The recommendation is that there are 5 seed examples in a seed examples file each with 3 question and answer pairs. Users may tweak these recommendations in the validator per their needs."
```
```suggestion
    "After assessment, the `qna.yaml` files can be quickly reviewed to ensure they includes the required elements and correct number of each. It is recommended to have 5 seed examples, each with 3 question and answer pairs. Users can adjust these recommendations in the validator per their needs."
```
Isnt this a requirement, not a recommendation?

kelbrown20 commented on notebooks/instructlab-knowledge/instructlab-knowledge.ipynb:
```diff
@@ -656,11 +641,11 @@
     "## Create Seed Dataset for SDG\n",
     "\n",
-    "This notebook combines the contents from the qna.yaml and the chunks from the source document to create a seed dataset for the synthetic data generation process.\n",
+    "This section combines the contents from the qna.yaml and the chunks from the source document to create a seed dataset for the synthetic data generation process.\n",
```
I think this description needs a refactor because I don't quite understand what is happening here

kelbrown20 commented on notebooks/instructlab-knowledge/instructlab-knowledge.ipynb:
```diff
@@ -727,21 +710,21 @@
-    "1. Converted the document using document and saved it to JSON for inspection\n",
+    "1. Converts the documents using Docling and saved it to JSON for inspection\n",
     "2. Split the extracted text into chunks\n",
     "3. Generated QA pairs for a subset of those chunks\n",
```
```suggestion
    "3. Generates Q&A pairs for a subset of those chunks\n",
```

kelbrown20 commented on notebooks/instructlab-knowledge/instructlab-knowledge.ipynb:
```diff
@@ -727,21 +710,21 @@
+    "To recap, given a source documents in PDF format, this notebook:\n",
     "\n",
-    "1. Converted the document using document and saved it to JSON for inspection\n",
+    "1. Converts the documents using Docling and saved it to JSON for inspection\n",
```
```suggestion
    "1. Converts the documents using Docling and saves it to JSON \n",
```

kelbrown20 commented on notebooks/instructlab-knowledge/instructlab-knowledge.ipynb:
```diff
@@ -727,21 +710,21 @@
     "\n",
-    "1. Converted the document using document and saved it to JSON for inspection\n",
+    "1. Converts the documents using Docling and saved it to JSON for inspection\n",
     "2. Split the extracted text into chunks\n",
```
```suggestion
    "2. Splits the JSON text into chunks\n",
```
^is this right?

kelbrown20 commented on notebooks/instructlab-knowledge/instructlab-knowledge.ipynb:
```diff
@@ -727,21 +710,21 @@
-    "4. Created a `qna.yaml` available for inspection and revision\n",
-    "5. Combined the chunks and `qna.yaml` to create a `seed.jsonl` for use with SDG\n",
+    "4. Creates a `qna.yaml` available for inspection and revision\n",
+    "5. Combined the chunks and `qna.yaml` to create a `seed_data.jsonl` for use with SDG\n",
```
```suggestion
    "5. Combines the chunks and `qna.yaml` to create a `seed_data.jsonl` to use for SDG\n",
```

kelbrown20 commented on notebooks/instructlab-knowledge/instructlab-knowledge.ipynb:
```diff
@@ -727,21 +710,21 @@
     "# Summary\n",
     "\n",
-    "To recap, given a source document in PDF format, this notebook:\n",
+    "To recap, given a source documents in PDF format, this notebook:\n",
```
Are you thinking of a specific tense for this, because before the tenses varied a lot and were inconsistent. It does matter to me which one, It should all just be in the same tense

alimaredia replied:
What would you suggest the tense we should stick with? Maybe we do a follow up PR aligning all of the tenses in all of the documentation within each section?

CC: @JustinXHale

iamemilio commented on notebooks/instructlab-knowledge/instructlab-knowledge.ipynb:
```diff
@@ -36,68 +130,130 @@
+   "outputs": [],
+   "source": [
+    "# Populated later on\n",
+    "contributions = []\n",
```
I like the idea of packaging these ups as "contributions". Since every `contribution` has the same set of properties, I think it would be clearer to use a class with structured fields in place of a dictionary here. It would significantly lower the margin of error, and make errors a lot easier to resolve.

iamemilio commented on notebooks/instructlab-knowledge/instructlab-knowledge.ipynb:
```diff
@@ -36,68 +130,130 @@
-    "    print(f\"***** Using {files[0]})\")\n",
+    "print(f\"Files to pre-process\\n--------------------\")\n",
+    "for contribution in contributions:\n",
+    "    print(f\"\\nContribution: {contribution[\"name\"]}\")\n",
+    "    print(f\"Files:\")\n",
```
```suggestion
    "    print(f\"\\nContribution: {contribution.get('name')}\")\n",
    "    print(\"Files:\")\n",
```
some nitpicks:
- using `.get` is safer since it return `None` if the key does not exist instead of raising an error

iamemilio commented:
> > I am starting to wonder if the `instructlab-knoweldge` notebook is doing too much, and I am not really sure why we are forcing it to be in a notebook, when it seems like what we want to do is to write a data conversion pipeline tool that wraps docling. If that's what we are delivering, wouldn't it be better off as a product, with standards, testing, explicit inputs and outputs, an sdk/cli/ui... etc, rather than an increasingly complex script that you run in a notebook?
> 
> @iamemilio I see where you're coming from, however I would lean against drawing boxes around any particular part of the workflow at the moment. Modularization is definitely the ultimate goal for a product, but from my understanding the initial push towards notebooks was so they could serve as a bucket of code where we could get an unbiased sense of all the customizability we would want. I think it makes sense to build this up to include all the customizability we would want and then break it down into more concrete modules once product requirements are clearer.

Yeah, I don't expect us to have any answers right now, especially since we are waiting for more direction from leadership. I'm just asking the question because that seems to be where the code is going, and as this script gets larger and more complex, it will get more difficult to maintain and use if we don't start applying some structure to it and defining parts of the user experience.

fabianofranz commented on notebooks/instructlab-knowledge/instructlab-knowledge.ipynb:
```diff
@@ -330,323 +510,128 @@
-   "metadata": {},
-   "source": [
-    "### Configure subset selection"
+    "API_KEY = os.getenv(\"MODEL_API_KEY\")  # the API access key for your account ( cannot be empty )\n",
+    "API_URL = os.getenv(\"MODEL_ENDPOINT_URL\")   # the URL of your model's API. Url can end in \"/v1\"\n",
+    "MODEL_ID = os.getenv(\"MODEL_NAME\") # the name of your model\n",
+    "NUM_SEED_EXAMPLES = 5"
```
+1, maybe something like this so we support env vars but preserve the capacity of setting them inline?

```Python
API_KEY = os.getenv("MODEL_API_KEY") or "none"  # the API access key for your account (cannot be empty)
API_URL = os.getenv("API_URL") or "http://127.0.0.1:11434/v1"   # the URL of your model's API. URL can end in "/v1"
MODEL_ID = os.getenv("MODEL_ID") or "granite3.3" # the name of your model
```

alimaredia replied:
@fabianofranz I like the way you structured it. An alternative is I just have commented out variable assignment with `os.getenv()` below the string assignment. 

I really like having the option in the notebook to not have to manually copy keys which was why I added it in the first place.

fabianofranz commented on .github/workflows/instructlab-knowledge-e2e.yml:
```diff
@@ -22,14 +22,15 @@ jobs:
               uses: actions/setup-python@v5
               with:
-                python-version: '3.11'
+                python-version: '3.12'
```
Just FYI, I successfully ran it end-to-end on 3.11 with just one minor adjustment. On notebooks/instructlab-knowledge/instructlab-knowledge.ipynb#232 , use `contribution['name']` (single quotes).

alimaredia commented on notebooks/instructlab-knowledge/instructlab-knowledge.ipynb:
```diff
@@ -7,20 +7,114 @@
+    "\n",
+    "`2023 Bylaws of organization Foo`\n",
+    "\n",
+    "Another example having of having multiple documents within the same contribution would be if the docuents had the same format. An example here could be grouping together a furniture company's instruction manuals. The format and layout of the instruction manuals would be the same accross different pieces of furniture, but each manual covers different furniture.\n",
```
Added to the latest changes.

alimaredia commented on notebooks/instructlab-knowledge/instructlab-knowledge.ipynb:
```diff
@@ -7,20 +7,114 @@
+    "\n",
+    "Usually contributions only have one document. Contributions with multiple documents happen when the subject matter and format are similar among a group of documents. \n",
+    "\n",
+    "An example of a contribution having multiple documents would the desire to teach a model an organization's bylaws over the years 2021, 2022, 2023, 2024, with a different PDF for each year.\n",
```
Added to the latest changes.

alimaredia commented on notebooks/instructlab-knowledge/instructlab-knowledge.ipynb:
```diff
@@ -284,7 +448,9 @@
-    "Next lets convert the document we want to chunk up into a Docling Document."
+    "Next lets convert the document we want to chunk up into a Docling Document.\n",
+    "\n",
+    "All chunks are saved to a JSON file called chunks.jsonl in the `chunks` directory in your contribution. This file is one of the inputs father below when we create the seed dataset for SDG."
```
I incorporated some of this into a revision of this sentence.

alimaredia commented:
@iamemilio @alinaryan @kelbrown20 @fabianofranz @Ryfernandes I've addressed most of your comments thanks for the feedback. To not keep increasing the scope of this PR there's 3 areas of work I'm going to open upstream issues on after this PR is merged:
1. Creating a `Contribution` data structure to enable all of the state of each contribution to be stored
2. Ensure that the example documents allow users to run through the notebook more consistently generates a proper qna.yaml
3. Do a docs review and ensure that there's consistency across the entire notebook

Ryfernandes commented:
Yeah that sounds great

khaledsulayman commented on notebooks/instructlab-knowledge/utils/create_seed_dataset.py:
```diff
@@ -19,12 +19,27 @@ def get_seed_dataset(path: str) -> Dataset:
     return ds
 
-def is_dir_valid(path: str) -> Path:
+def is_dir_valid(path: str) -> None:
```
nit:
```suggestion
def validate_dir(path: str) -> Path:
```
not sure why the return signature was changed from `Path` to `None` here, but probably could use some renaming

khaledsulayman commented on notebooks/instructlab-knowledge/instructlab-knowledge.ipynb:
```diff
@@ -7,20 +7,114 @@
+   "id": "0b02a66e-125e-47e6-9b6b-5f49d50990ca",
+   "metadata": {},
+   "source": [
+    "### Getting Started\n",
+    "\n",
+    "The first step in this notebook is to establish a workspace. Workspaces allow multi-tenancy or multiple different runs of this notebook. Without workspaces the results of each of the steps would be overwritten each time this notebook is executed.\n",
+    "\n",
+    "Users should change the `WORKSPACE_NAME` to suite their needs."
+   ]
+  },
```
I feel like we could use something more explicitly stating that _regardless of whatever subset of this notebook the user wants to run, they should first run the two code cells following this,_ since this is what sets up the given workspace or verifies that the directories exist as expected.

khaledsulayman reviewed (commented):
left 2 minor comments. Otherwise, everything looks functionally good. Thanks!

alimaredia commented:
@khaledsulayman Thanks for the comments! The function you pointed out was deleted since it's not being used anymore and I made some adjustments to the sections in the notebook and the documentation to make it easier to navigate back to the first two cells to re-run them. I was constantly doing that too.

iamemilio reviewed (approved):
+1 to follow up tasks. LGTM
